cli/lsp/repl.rs

```diff
@@ -0,0 +1,308 @@
+              uri: self.get_document_specifier(),
+            },
+            position: Position {
+              line: line_and_column.line_index as u32,
+              character: line_and_column.column_index as u32,
+            },
+          },
+          work_done_progress_params: WorkDoneProgressParams {
+            work_done_token: None,
+          },
+          partial_result_params: PartialResultParams {
+            partial_result_token: None,
+          },
+          context: None,
+        })
+        .await
+        .ok()
+        .unwrap_or_default();
+
+      let items = match response {
+        Some(CompletionResponse::Array(items)) => items,
+        Some(CompletionResponse::List(list)) => list.items,
+        None => Vec::new(),
+      };
+      items
+        .into_iter()
+        .filter_map(|item| {
+          item.text_edit.and_then(|edit| match edit {
+            CompletionTextEdit::Edit(edit) => Some(ReplCompletionItem {
+              new_text: edit.new_text,
+              span: lsp_range_to_span(&text_info, &edit.range),
+            }),
+            CompletionTextEdit::InsertAndReplace(_) => None,
+          })
+        })
+        .filter(|item| {
+          // filter the results to only exact matches
+          let text = &text_info.text_str()
+            [item.span.lo.0 as usize..item.span.hi.0 as usize];
+          item.new_text.starts_with(text)
+        })
+        .map(|mut item| {
+          // convert back to a line position
+          item.span = Span::new(
+            item.span.lo - before_line_len,
+            item.span.hi - before_line_len,
+            Default::default(),
+          );
+          item
+        })
+        .collect()
+    })
+    .await
+  }
+
+  async fn did_change(&mut self, new_text: &str) {
+    self.check_cwd_change().await;
+    let new_text = if new_text.ends_with('\n') {
+      new_text.to_string()
+    } else {
+      format!("{}\n", new_text)
+    };
+    self.document_version += 1;
+    let current_line_count =
+      self.document_text.chars().filter(|c| *c == '\n').count() as u32;
+    let pending_line_count =
+      self.pending_text.chars().filter(|c| *c == '\n').count() as u32;
+    self
+      .language_server
+      .did_change(DidChangeTextDocumentParams {
+        text_document: VersionedTextDocumentIdentifier {
+          uri: self.get_document_specifier(),
+          version: self.document_version,
+        },
+        content_changes: vec![TextDocumentContentChangeEvent {
+          range: Some(Range {
+            start: Position::new(current_line_count, 0),
+            end: Position::new(current_line_count + pending_line_count, 0),
+          }),
+          range_length: None,
+          text: new_text.to_string(),
+        }],
+      })
+      .await;
+    self.pending_text = new_text;
+  }
+
+  async fn check_cwd_change(&mut self) {
+    // handle if the cwd changes
+    let cwd_uri = get_cwd_uri().unwrap();
```

Comment:
You should probably handle the error instead of panicking. Deleting the working directory is not an unheard of thing.

Reply:
I updated the code to handle this, but I'm not sure it's the best way of handling it though.